src/backend/distributed/commands/foreign_constraint.c

```diff
@@ -696,89 +617,159 @@ HeapTupleOfForeignConstraintIncludesColumn(HeapTuple heapTuple, Oid relationId,
+	return list_length(tableReferencingForeignConstraintOids) > 0;
+}
 
-	ScanKeyInit(&scanKey[0], Anum_pg_constraint_conrelid, BTEqualStrategyNumber, F_OIDEQ,
-				relationId);
-	SysScanDesc scanDescriptor = systable_beginscan(pgConstraint, scanIndexId, useIndex,
-													NULL,
-													scanKeyCount, scanKey);
 
-	HeapTuple heapTuple = systable_getnext(scanDescriptor);
-	while (HeapTupleIsValid(heapTuple))
+/*
+ * ConstraintIsAForeignKey returns true if the given constraint name
+ * is a foreign key defined on the relation.
+ */
+bool
+ConstraintIsAForeignKey(char *inputConstaintName, Oid relationOid)
+{
+	int flags = EXTRACT_REFERENCING_CONSTRAINTS;
+	List *foreignKeyOids = GetTableForeignConstraintOidsInternal(relationOid, flags);
+
+	Oid foreignKeyOid = FindForeignKeyOidWithName(foreignKeyOids, inputConstaintName);
+
+	return OidIsValid(foreignKeyOid);
+}
+
+
+/*
+ * FindForeignKeyOidWithName searches the foreign key constraint with
+ * inputConstraintName in the given list of foreign key constraint OIDs.
+ * Returns the OID of the matching constraint. If there no matching constraint
+ * in the given list, then returns InvalidOid.
+ */
+static Oid
+FindForeignKeyOidWithName(List *foreignKeyOids, const char *inputConstraintName)
+{
+	Oid foreignKeyOid = InvalidOid;
+	foreach_oid(foreignKeyOid, foreignKeyOids)
 	{
-		Form_pg_constraint constraintForm = (Form_pg_constraint) GETSTRUCT(heapTuple);
+		char *constraintName = get_constraint_name(foreignKeyOid);
 
-		if (constraintForm->contype == CONSTRAINT_FOREIGN)
+		if (strncmp(constraintName, inputConstraintName, NAMEDATALEN) == 0)
 		{
-			systable_endscan(scanDescriptor);
-			heap_close(pgConstraint, NoLock);
-
-			return true;
+			return foreignKeyOid;
 		}
-
-		heapTuple = systable_getnext(scanDescriptor);
 	}
 
-	systable_endscan(scanDescriptor);
-	heap_close(pgConstraint, NoLock);
-
-	return false;
+	return InvalidOid;
 }
 
 
 /*
- * ConstraintIsAForeignKey returns true if the given constraint name
- * is a foreign key to defined on the relation.
+ * GetTableForeignConstraintOidsInternal takes in a relationOid, and returns
+ * the list of OIDs for foreign constraints that the relation with relationOid
+ * is involved according to "flags" argument. In that sense,
+ * "flags" should include either:
+ *  - "EXTRACT_REFERENCING_CONSTRAINTS" to get the ones in which our relation is
+ *    the "referencing" one, or
+ *  - "EXTRACT_REFERENCED_CONSTRAINTS" to get the ones in which our relation is
+ *    the "referenced" one.
+ * Self-referencing foreign keys will not be considered if "flags" includes
+ * EXCLUDE_SELF_REFERENCES as well.
  */
-bool
-ConstraintIsAForeignKey(char *constraintNameInput, Oid relationId)
+static List *
+GetTableForeignConstraintOidsInternal(Oid relationOid, int flags)
 {
+	AttrNumber attributeNumber = InvalidAttrNumber;
+
+	/* either of them should be passed */
+	Assert(flags & (EXTRACT_REFERENCING_CONSTRAINTS | EXTRACT_REFERENCED_CONSTRAINTS));
```

Comment:
If we want only one of them to be passed at a time we can do an xor here and remove the Assert below.

Using xor would return true only if one of them is set:
1 xor 1 = 0
1 xor 0 = 1
0 xor 1 = 1
0 xor 0 = 1

Assert( (flags & EXTRACT_REFERENCING_CONSTRAINTS) ^ (flags &  EXTRACT_REFERENCED_CONSTRAINTS))

Reply:
Why do we need this Assert? I mean, why can't we  allow both?

Reply:
(I can see why, we should add a clarifying comment)

Reply:
> Using xor would return true only if one of them is set:

I think not everyone is as smart as you to recognize this immediately :) So, can we use something simpler?

!(flags & EXTRACT_REFERENCING_CONSTRAINTS && flags & EXTRACT_REFERENCED_CONSTRAINTS)